hoomd/md/EvaluatorPairFourier.h

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2009-2016 The Regents of the University of Michigan
+// This file is part of the HOOMD-blue project, released under the BSD 3-Clause License.
+
+// Maintainer: zhoupj
+
+#ifndef __PAIR_EVALUATOR_FOURIER_H__
+#define __PAIR_EVALUATOR_FOURIER_H__
+
+#ifndef NVCC
+#include <string>
+#endif
+
+#include "hoomd/HOOMDMath.h"
+
+/*! \file EvaluatorPairFourier.h
+    \brief Defines the pair evaluator class for potential in form of Fourier series
+
+    \details .....
+*/
+
+// need to declare these class methods with __device__ qualifiers when building in nvcc
+// DEVICE is __host__ __device__ when included in nvcc and blank when included into the host compiler
+#ifdef NVCC
+#define DEVICE __device__
+#else
+#define DEVICE
+#endif
+
+// TODO: documentation
+//! Class for evaluating the Fourier pair potential
+/*! <b>General Overview</b>
+
+    See EvaluatorPairLJ
+
+    <b>Fourier specifics</b>
+
+    EvaluatorPairFourier evaluates the function:
+
+
+    The Fourier potential does not need diameter or charge. 18 parameters are specified and stored in a two array for a and b.
+    \a a is placed in \a params.a, \a b is in \a params.b.
+
+*/
+struct pair_fourier_params
+{
+  Scalar a[3];      //!< Fourier component coefficents
+  Scalar b[3];      //!< Fourier component coefficents
+};
+
+//! Function to make the parameter type
+DEVICE inline pair_fourier_params make_pair_fourier_params(pybind11::list a, pybind11::list b)
+    {
+    pair_fourier_params retval;
+    for (int i = 0; i < 3; ++i)
+        {
+        retval.a[i] = pybind11::cast<Scalar>(a[i]);
+        retval.b[i] = pybind11::cast<Scalar>(b[i]);
+        }
+    return retval;
+    }
+
+
+class EvaluatorPairFourier
+    {
+    public:
+        //! Define the parameter type used by this pair potential evaluator
+        typedef pair_fourier_params param_type; //first try a 4th order fourier expression of potential
+        //! Constructs the pair potential evaluator
+        /*! \param _rsq Squared distance beteen the particles
+            \param _rcutsq Sqauared distance at which the potential goes to 0
+            \param _params Per type pair parameters of this potential
+        */
+        DEVICE EvaluatorPairFourier(Scalar _rsq, Scalar _rcutsq,
+                                    const param_type& _params)
+            {
```

Comment:
The compiler might not optimize out the copy into `a[]` and `b[]`. I recommend declaring a class member variable `const pair_fourier_params& params`, initialized with `: params(_params)` here.